### ISSUE TITLE
Small object detection

### ISSUE THREAD

**Issue by CesareDavidePace** [ultralytics/yolov5#10087]
### Discussed in https://github.com/ultralytics/yolov5/discussions/10083

<div type='discussions-op-text'>

<sup>Originally posted by **CesareDavidePace** November  8, 2022</sup>
Hi, 
I would need help with regards to the detection of very small objects, we are talking about objects even 20x20 on images of 5000x3000.
My approach was to use the sahi framework to do sliding windows and to keep the native size. In particular, my dataset is made up of satellite images.
My questions are:
1) What model can I use to improve the detection of very small objects? I have seen many recommend the yoloP2 but I don't know what changes.
2) Is it worth doing pretraining with the coco or training from scratch? And are there models already trained for images from above?
3) I did sliding windows in 500x500, is it better to increase the tiles for example 1280x1280 since the networks are designed for that?</div>

**Comment by glenn-jocher** [comment]
@CesareDavidePace I'd simply train and run inference at native resolution (or the largest resolution that you can). 20 pix is not a small object, it's a medium size object. Small stride P3 output is 8 pixels.

**Comment by CesareDavidePace** [comment]
so you are telling me i train the net with 5000x300 images without tiling, right?

**Comment by glenn-jocher** [comment]
That's what I would do, or the largest that you can fit with --batch-size 1

**Comment by CesareDavidePace** [comment]
okay, I'll try. and what about the model? which one should I choose?

**Comment by glenn-jocher** [comment]
Model selection is up to you and depends on factors beyond my understanding like resource availability, response time and accuracy requirement, to name just a few.

If you want a comparison of model performance on COCO see README.

**Comment by CesareDavidePace** [comment]
Perfect, thanks

**Comment by myasser63** [comment]
@glenn-jocher Can I train YOLOv5s on bigger image size than the default 640x640 or it would affect the results.

**Comment by CesareDavidePace** [comment]
@glenn-jocher i tried as you told me then full resolution and i got 0 as mAP

**Comment by CesareDavidePace** [comment]
![labels](https://user-images.githubusercontent.com/104144011/200814934-415f0ec2-6cff-402b-b4bb-a3f6044c973e.jpg)

**Comment by glenn-jocher** [comment]
👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Label verification.** View `train_batch*.jpg` on train start to verify your labels appear correct, i.e. see [example](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data#local-logging) mosaic.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch-low.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyps/hyp.scratch-low.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by CesareDavidePace** [comment]
Hello,
I tried as you advised me. That is, I used the highest possible resolution and 
practically what I get is that during training the performances do not converge.
![Cattura](https://user-images.githubusercontent.com/104144011/201485612-17db3d3f-6a6a-4a65-87b4-36d29c5739fe.PNG)


these are the hyp used:
hyp:
  lr0: 0.01
  lrf: 0.1
  momentum: 0.937
  weight_decay: 0.0005
  warmup_epochs: 3.0
  warmup_momentum: 0.8
  warmup_bias_lr: 0.1
  box: 0.05
  cls: 0.0
  cls_pw: 0.0
  obj: 0.7
  obj_pw: 1.0
  iou_t: 0.2
  anchor_t: 4.0
  fl_gamma: 0.0
  hsv_h: 0.015
  hsv_s: 0.7
  hsv_v: 0.4
  degrees: 0.1
  translate: 0.1
  scale: 1
  shear: 0.0
  perspective: 0.0
  flipud: 0.0
  fliplr: 0.5
  mosaic: 1.0
  mixup: 0.1
  copy_paste: 0.0

**Comment by dhrubapuc23** [comment]
> Hello @glenn-jocher , I tried as you advised me. That is, I used the highest possible resolution and practically what I get is that during training the performances do not converge. ![Cattura](https://user-images.githubusercontent.com/104144011/201485612-17db3d3f-6a6a-4a65-87b4-36d29c5739fe.PNG)
> 
> these are the hyp used: hyp: lr0: 0.01 lrf: 0.1 momentum: 0.937 weight_decay: 0.0005 warmup_epochs: 3.0 warmup_momentum: 0.8 warmup_bias_lr: 0.1 box: 0.05 cls: 0.0 cls_pw: 0.0 obj: 0.7 obj_pw: 1.0 iou_t: 0.2 anchor_t: 4.0 fl_gamma: 0.0 hsv_h: 0.015 hsv_s: 0.7 hsv_v: 0.4 degrees: 0.1 translate: 0.1 scale: 1 shear: 0.0 perspective: 0.0 flipud: 0.0 fliplr: 0.5 mosaic: 1.0 mixup: 0.1 copy_paste: 0.0

Try with different lr0

**Comment by glenn-jocher** [comment]
@dhrubapuc23 if you are seeing that your model training performances do not converge, it could be due to the learning rate being too high, too low, or learning rate scheduling. 

I suggest trying a different initial learning rate `lr0` value for your training. Try decreasing the `lr0` value in half, and if the performance still doesn't converge after ~10 epochs, you can decrease the `lr0` by another 50% and continue until you see some convergence. 

Moreover, to achieve great performance you should try experimenting with different LR schedules. For example, you may want to use learning rate schedulers (e.g. `OneCycleLR`, `CosineAnnealingLR`) to reduce the LR as you train.

I hope this helps! Let me know if there's anything else I can assist you with.

**Comment by dhrubapuc23** [comment]
> @dhrubapuc23 if you are seeing that your model training performances do not converge, it could be due to the learning rate being too high, too low, or learning rate scheduling.
> 
> I suggest trying a different initial learning rate `lr0` value for your training. Try decreasing the `lr0` value in half, and if the performance still doesn't converge after ~10 epochs, you can decrease the `lr0` by another 50% and continue until you see some convergence.
> 
> Moreover, to achieve great performance you should try experimenting with different LR schedules. For example, you may want to use learning rate schedulers (e.g. `OneCycleLR`, `CosineAnnealingLR`) to reduce the LR as you train.
> 
> I hope this helps! Let me know if there's anything else I can assist you with.

Thanks for the suggestion.

**Comment by glenn-jocher** [comment]
@dhrubapuc23 you're welcome! Good luck with your training and feel free to reach out if you have any further questions.

**Comment by aash1999** [comment]
@dhrubapuc23  You can check this Repo for small object detection using YoloV5 : https://github.com/aash1999/yolov5-cbam

**Comment by star4s** [comment]
@dhrubapuc23 @glenn-jocher I dropped by the link, [https://github.com/aash1999/yolov5-cbam](url)
CBAM was constructed based on yolo5m.yaml, which defines the model structure of yolov5.
Could you please share the file that configures CBAM based on yolov5x6.yaml?

**Comment by glenn-jocher** [comment]
@star4s hello! It's great to see your interest in adapting CBAM to different YOLOv5 models. Unfortunately, I don't have a specific `yolov5x6.yaml` file configured for CBAM. Typically, integrating CBAM into a different model architecture involves modifying the model's YAML file to include CBAM layers at appropriate points in the network.

You might need to manually adjust the `yolov5x6.yaml` based on the modifications seen in `yolo5m.yaml` from the CBAM repository. This usually involves adding CBAM blocks after each convolutional layer or wherever you see fit based on your understanding of the model's architecture and your specific needs.

If you're not familiar with how to do this, I recommend studying the changes made in the `yolo5m.yaml` and trying to replicate them in the `yolov5x6.yaml`. Experimenting and testing are key here.

Good luck, and let us know how it goes!
